diff -rubN phylip-3.697/src/phylip.h phylip-3.697-new/src/phylip.h
--- phylip-3.697/src/phylip.h	2017-12-07 20:19:20.000000000 -0500
+++ phylip-3.697-new/src/phylip.h	2022-07-24 23:31:37.629634312 -0400
@@ -265,8 +265,8 @@
 /*** Static memory parameters ***/
 
 #define FNMLNGTH        200  /* length of array to store a file name */
-#define nmlngth         10   /* number of characters in species name    */
-#define MAXNCH          20   /* must be greater than or equal to nmlngth */
+#define nmlngth         200   /* number of characters in species name    */
+#define MAXNCH          200   /* must be greater than or equal to nmlngth */
 #define maxcategs       9    /* maximum number of site types */
 #define maxcategs2     11    /* maximum number of site types + 2 */
 #define point           "."
@@ -331,7 +331,7 @@
 /* Lower-triangular format. */
 #define MAT_LOWERTRI    (MAT_LOWER | MAT_MACHINE)
 
-boolean javarun;
+static boolean javarun;
 
 typedef long *steptr;
 typedef long longer[6];
@@ -351,7 +351,7 @@
 extern long spp, words, bits;
 extern boolean ibmpc, ansi, tranvsp;
 extern naym *nayme;                     /* names of species */
-boolean firstplotblock; // for debugging BMP output
+static boolean firstplotblock; // for debugging BMP output
 
 #define ebcdic          EBCDIC
 
diff -rubN phylip-3.697/src/seq.h phylip-3.697-new/src/seq.h
--- phylip-3.697/src/seq.h	2014-09-19 11:42:32.000000000 -0400
+++ phylip-3.697-new/src/seq.h	2022-07-24 23:31:50.749632041 -0400
@@ -37,7 +37,7 @@
 #define SEQ_H
 
 #define ebcdic          EBCDIC
-#define MAXNCH          20
+#define MAXNCH          200
 
 /* All of this came over from cons.h    -plc*/ 
 #define OVER              7
